sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,635 @@
+          + "query");
+      return builder().setQuery(query).build();
+    }
+
+    public Read withScrollKeepalive(String scrollKeepalive) {
+      checkArgument(scrollKeepalive != null, "ElasticsearchIO.read().withScrollKeepalive"
+          + "(keepalive) called with null keepalive");
+      return builder().setScrollKeepalive(scrollKeepalive).build();
+    }
+
+    @Override
+    public PCollection<String> apply(PBegin input) {
+      return input.apply(org.apache.beam.sdk.io.Read.from(
+          new BoundedElasticsearchSource(this, null)
+      ));
+    }
+
+    @Override
+    public void validate(PBegin input) {
+      checkState(getConnectionConfiguration() != null, "ElasticsearchIO.read() requires a "
+          + "connection configuration to be set via withConnectionConfiguration(configuration)");
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      builder.addIfNotNull(DisplayData.item("query", getQuery()));
+      getConnectionConfiguration().populateDisplayData(builder);
+    }
+
+  }
+
+  /**
+   * A {@link BoundedSource} reading from Elasticsearch.
+   */
+  protected static class BoundedElasticsearchSource extends BoundedSource<String> {
+
+    private final ElasticsearchIO.Read spec;
+    // shardPreference is the shard number where the source will read the documents
+    @Nullable private final String shardPreference;
+
+    protected BoundedElasticsearchSource(Read spec, String shardPreference) {
+      this.spec = spec;
+      this.shardPreference = shardPreference;
+    }
+
+    @Override
+    public List<? extends BoundedSource<String>> splitIntoBundles(long desiredBundleSizeBytes,
+                                                                  PipelineOptions options)
+        throws Exception {
+      List<BoundedElasticsearchSource> sources = new ArrayList<>();
+
+/*
+       1.We split per shard :
+       unfortunately, Elasticsearch 2.x doesn't provide a way to do parallel reads on a single
+       shard. So  we do not use desiredBundleSize because we cannot split shards.
+       With the slice API in ES 5.0 we will be able to use desiredBundleSize.
+       Basically we will just ask the slice API to return data
+       in nbBundles = estimatedSize / desiredBundleSize chuncks.
+       So each beam source will read around desiredBundleSize volume of data.
+
+       2.We read only primary shards
+       let's say the ES index is configured to split data into 5 shards with a replica of 1
+       (for failover). In that case, there will be a total of 10 shards (5 primary, 5 replica).
+       Each beam source reads the documents contained in one shard.
+       As we don't want to have duplicate documents in PCollection, we only read primary shards.
+       In case of failover (ES side), primary shards stored in a failing ES node
+       are no longer available and replica shards (stored in other nodes) become primary shards.
```

Comment:
I just talked to one guy of elasticsearch dev team. In fact, my comment is partially wrong: primary and replica shards have the same shard_id. So the even if we specify preference=shards:2, ES load balances (round robin) the request between primary shard 2 and replica shard 2. But, as each shard is responsible for only one part of the data, there will be no duplicate.